###Heart Disease Prediction


Heart disease is a general term that includes many types of heart problems. It's also called cardiovascular disease, which means heart and blood vessel disease. 
The causes of heart disease depend on the type of disease. Some possible causes include lifestyle, genetics, infections, medicines, and other diseases.



This classification model will predict whether the patient has heart disease or not based on various conditions/symptoms of their body.

In [0]:
%scala 

// File location and type
val file_location = "/FileStore/tables/heart.csv"
val file_type = "csv"

// options
val infer_schema = "true"
val first_row_is_header = "true"
val delimiter = ","
// create a df
val heartDF = spark.read.format(file_type) 
  .option("inferSchema", infer_schema) 
  .option("header", first_row_is_header) 
  .option("sep", delimiter) 
  .load(file_location)

heartDF.show()

+---+---+---+--------+----+---+-------+-------+-----+-------+-----+---+----+------+
age|sex| cp|trestbps|chol|fbs|restecg|thalach|exang|oldpeak|slope| ca|thal|target|
+---+---+---+--------+----+---+-------+-------+-----+-------+-----+---+----+------+
 63| 1| 3| 145| 233| 1| 0| 150| 0| 2.3| 0| 0| 1| 1|
 37| 1| 2| 130| 250| 0| 1| 187| 0| 3.5| 0| 0| 2| 1|
 41| 0| 1| 130| 204| 0| 0| 172| 0| 1.4| 2| 0| 2| 1|
 56| 1| 1| 120| 236| 0| 1| 178| 0| 0.8| 2| 0| 2| 1|
 57| 0| 0| 120| 354| 0| 1| 163| 1| 0.6| 2| 0| 2| 1|
 57| 1| 0| 140| 192| 0| 1| 148| 0| 0.4| 1| 0| 1| 1|
 56| 0| 1| 140| 294| 0| 0| 153| 0| 1.3| 1| 0| 2| 1|
 44| 1| 1| 120| 263| 0| 1| 173| 0| 0.0| 2| 0| 3| 1|
 52| 1| 2| 172| 199| 1| 1| 162| 0| 0.5| 2| 0| 3| 1|
 57| 1| 2| 150| 168| 0| 1| 174| 0| 1.6| 2| 0| 2| 1|
 54| 1| 0| 140| 239| 0| 1| 160| 0| 1.2| 2| 0| 2| 1|
 48| 0| 2| 130| 275| 0| 1| 139| 0| 0.2| 2| 0| 2| 1|
 49| 1| 1| 130| 266| 0| 1| 171| 0| 0.6| 2| 0| 2| 1|
 64| 1| 3| 110| 211| 0| 0| 144| 1| 1.8| 1| 0| 2| 1|
 58| 0| 3| 150| 283| 1| 0| 162| 0| 1.0| 2| 0| 2| 1|
 50| 0| 2| 120| 219| 0| 1| 158| 0| 1.6| 1| 0| 2| 1|
 58| 0| 2| 120| 340| 0| 1| 172| 0| 0.0| 2| 0| 2| 1|
 66| 0| 3| 150| 226| 0| 1| 114| 0| 2.6| 0| 0| 2| 1|
 43| 1| 0| 150| 247| 0| 1| 171| 0| 1.5| 2| 0| 2| 1|
 69| 0| 3| 140| 239| 0| 1| 151| 0| 1.8| 2| 2| 2| 1|
+---+---+---+--------+----+---+-------+-------+-----+-------+-----+---+----+------+
only showing top 20 rows

file_location: String = /FileStore/tables/heart.csv
file_type: String = csv
infer_schema: String = true
first_row_is_header: String = true
delimiter: String = ,
heartDF: org.apache.spark.sql.DataFrame = [age: int, sex: int ... 12 more fields]

* **age:** The person's age in years

* **sex:** The person's sex (1 = male, 0 = female)

* **cp:** The chest pain experienced (Value 1: typical angina, Value 2: atypical angina, Value 3: non-anginal pain, Value 4: asymptomatic)

* **trestbps:** The person's resting blood pressure (mm Hg on admission to the hospital)

* **chol:** The person's cholesterol measurement in mg/dl

* **fbs:** The person's fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)

* **restecg:** Resting electrocardiographic measurement (0 = normal, 1 = having ST-T wave abnormality, 2 = showing probable or definite left ventricular hypertrophy by Estes' 
criteria)

* **thalach:** The person's maximum heart rate achieved

* **exang:** Exercise induced angina (1 = yes; 0 = no)

* **oldpeak:** ST depression induced by exercise relative to rest ('ST' relates to positions on the ECG plot. See more here)

* **slope:** the slope of the peak exercise ST segment (Value 1: upsloping, Value 2: flat, Value 3: downsloping)

* **ca:** The number of major vessels (0-3)

* **thal:** A blood disorder called thalassemia

* **target:** Heart disease (0 = no, 1 = yes)

#### Diagnosis: 
The diagnosis of heart disease is done on a combination of clinical signs and test results.
<br>
 Examples - 
* From electrocardiograms and cardiac computerized tomography (CT) scans, to blood tests and exercise stress tests.
* High cholesterol, high blood pressure, diabetes, weight, family history and smoking.
* Hereditary variables such as Thalassemia.
* Smoking, high cholesterol, high blood pressure, physical inactivity, being overweight and having diabetes.
* Stress, alcohol and poor diet/nutrition.

<br>
Hypothesis - if the above model happens to be predictive, we'll see some interesting factors standing out as compared to the others.

In [0]:
%scala

heartDF.count()

res2: Long = 303

In [0]:
%scala

heartDF.select("age", "sex", "cp", "trestbps", "chol", "fbs","restecg", "thalach","exang","oldpeak","slope","ca","thal","target").describe().show()


+-------+------------------+-------------------+------------------+------------------+------------------+-------------------+-----------------+------------------+-------------------+------------------+------------------+------------------+------------------+------------------+
summary| age| sex| cp| trestbps| chol| fbs| restecg| thalach| exang| oldpeak| slope| ca| thal| target|
+-------+------------------+-------------------+------------------+------------------+------------------+-------------------+-----------------+------------------+-------------------+------------------+------------------+------------------+------------------+------------------+
 count| 303| 303| 303| 303| 303| 303| 303| 303| 303| 303| 303| 303| 303| 303|
 mean|54.366336633663366| 0.6831683168316832| 0.966996699669967|131.62376237623764|246.26402640264027| 0.1485148514851485|0.528052805280528|149.64686468646866|0.32673267326732675|1.0396039603960396|1.3993399339933994|0.7293729372937293|2.3135313531353137|0.5445544554455446|
 stddev| 9.08210098983786|0.46601082333962385|1.0320524894832983| 17.5381428135171| 51.83075098793005|0.35619787492797644|0.525859596359298| 22.90516111491409|0.46979446452231655|1.1610750220686346|0.6162261453459622|1.0226063649693276|0.6122765072781408|0.4988347841643915|
 min| 29| 0| 0| 94| 126| 0| 0| 71| 0| 0.0| 0| 0| 0| 0|
 max| 77| 1| 3| 200| 564| 1| 2| 202| 1| 6.2| 2| 4| 3| 1|
+-------+------------------+-------------------+------------------+------------------+------------------+-------------------+-----------------+------------------+-------------------+------------------+------------------+------------------+------------------+------------------+

In [0]:
%scala

heartDF.select("age", "sex", "cp", "chol").describe().show()


+-------+------------------+-------------------+------------------+------------------+
summary| age| sex| cp| chol|
+-------+------------------+-------------------+------------------+------------------+
 count| 303| 303| 303| 303|
 mean|54.366336633663366| 0.6831683168316832| 0.966996699669967|246.26402640264027|
 stddev| 9.08210098983786|0.46601082333962385|1.0320524894832983| 51.83075098793005|
 min| 29| 0| 0| 126|
 max| 77| 1| 3| 564|
+-------+------------------+-------------------+------------------+------------------+

In [0]:
%scala

heartDF.printSchema()

root
-- age: integer (nullable = true)
-- sex: integer (nullable = true)
-- cp: integer (nullable = true)
-- trestbps: integer (nullable = true)
-- chol: integer (nullable = true)
-- fbs: integer (nullable = true)
-- restecg: integer (nullable = true)
-- thalach: integer (nullable = true)
-- exang: integer (nullable = true)
-- oldpeak: double (nullable = true)
-- slope: integer (nullable = true)
-- ca: integer (nullable = true)
-- thal: integer (nullable = true)
-- target: integer (nullable = true)

In [0]:
%scala

heartDF.createOrReplaceTempView("HeartData");

In [0]:
%sql

select * from HeartData;

age sex cp trestbps chol fbs restecg thalach exang oldpeak slope ca thal target 63 1 3 145 233 1 0 150 0 2.3 0 0 1 1 37 1 2 130 250 0 1 187 0 3.5 0 0 2 1 41 0 1 130 204 0 0 172 0 1.4 2 0 2 1 56 1 1 120 236 0 1 178 0 0.8 2 0 2 1 57 0 0 120 354 0 1 163 1 0.6 2 0 2 1 57 1 0 140 192 0 1 148 0 0.4 1 0 1 1 56 0 1 140 294 0 0 153 0 1.3 1 0 2 1 44 1 1 120 263 0 1 173 0 0.0 2 0 3 1 52 1 2 172 199 1 1 162 0 0.5 2 0 3 1 57 1 2 150 168 0 1 174 0 1.6 2 0 2 1 54 1 0 140 239 0 1 160 0 1.2 2 0 2 1 48 0 2 130 275 0 1 139 0 0.2 2 0 2 1 49 1 1 130 266 0 1 171 0 0.6 2 0 2 1 64 1 3 110 211 0 0 144 1 1.8 1 0 2 1 58 0 3 150 283 1 0 162 0 1.0 2 0 2 1 50 0 2 120 219 0 1 158 0 1.6 1 0 2 1 58 0 2 120 340 0 1 172 0 0.0 2 0 2 1 66 0 3 150 226 0 1 114 0 2.6 0 0 2 1 43 1 0 150 247 0 1 171 0 1.5 2 0 2 1 69 0 3 140 239 0 1 151 0 1.8 2 2 2 1 59 1 0 135 234 0 1 161 0 0.5 1 0 3 1 44 1 2 130 233 0 1 179 1 0.4 2 0 2 1 42 1 0 140 226 0 1 178 0 0.0 2 0 2 1 61 1 2 150 243 1 1 137 1 1.0 1 0 2 1 40 1 3 140 199 0 1 178 1 1.4 2 0 3 1 71 0 1 160 302 0 1 162 0 0.4 2 2 2 1 59 1 2 150 212 1 1 157 0 1.6 2 0 2 1 51 1 2 110 175 0 1 123 0 0.6 2 0 2 1 65 0 2 140 417 1 0 157 0 0.8 2 1 2 1 53 1 2 130 197 1 0 152 0 1.2 0 0 2 1 41 0 1 105 198 0 1 168 0 0.0 2 1 2 1 65 1 0 120 177 0 1 140 0 0.4 2 0 3 1 44 1 1 130 219 0 0 188 0 0.0 2 0 2 1 54 1 2 125 273 0 0 152 0 0.5 0 1 2 1 51 1 3 125 213 0 0 125 1 1.4 2 1 2 1 46 0 2 142 177 0 0 160 1 1.4 0 0 2 1 54 0 2 135 304 1 1 170 0 0.0 2 0 2 1 54 1 2 150 232 0 0 165 0 1.6 2 0 3 1 65 0 2 155 269 0 1 148 0 0.8 2 0 2 1 65 0 2 160 360 0 0 151 0 0.8 2 0 2 1 51 0 2 140 308 0 0 142 0 1.5 2 1 2 1 48 1 1 130 245 0 0 180 0 0.2 1 0 2 1 45 1 0 104 208 0 0 148 1 3.0 1 0 2 1 53 0 0 130 264 0 0 143 0 0.4 1 0 2 1 39 1 2 140 321 0 0 182 0 0.0 2 0 2 1 52 1 1 120 325 0 1 172 0 0.2 2 0 2 1 44 1 2 140 235 0 0 180 0 0.0 2 0 2 1 47 1 2 138 257 0 0 156 0 0.0 2 0 2 1 53 0 2 128 216 0 0 115 0 0.0 2 0 0 1 53 0 0 138 234 0 0 160 0 0.0 2 0 2 1 51 0 2 130 256 0 0 149 0 0.5 2 0 2 1 66 1 0 120 302 0 0 151 0 0.4 1 0 2 1 62 1 2 130 231 0 1 146 0 1.8 1 3 3 1 44 0 2 108 141 0 1 175 0 0.6 1 0 2 1 63 0 2 135 252 0 0 172 0 0.0 2 0 2 1 52 1 1 134 201 0 1 158 0 0.8 2 1 2 1 48 1 0 122 222 0 0 186 0 0.0 2 0 2 1 45 1 0 115 260 0 0 185 0 0.0 2 0 2 1 34 1 3 118 182 0 0 174 0 0.0 2 0 2 1 57 0 0 128 303 0 0 159 0 0.0 2 1 2 1 71 0 2 110 265 1 0 130 0 0.0 2 1 2 1 54 1 1 108 309 0 1 156 0 0.0 2 0 3 1 52 1 3 118 186 0 0 190 0 0.0 1 0 1 1 41 1 1 135 203 0 1 132 0 0.0 1 0 1 1 58 1 2 140 211 1 0 165 0 0.0 2 0 2 1 35 0 0 138 183 0 1 182 0 1.4 2 0 2 1 51 1 2 100 222 0 1 143 1 1.2 1 0 2 1 45 0 1 130 234 0 0 175 0 0.6 1 0 2 1 44 1 1 120 220 0 1 170 0 0.0 2 0 2 1 62 0 0 124 209 0 1 163 0 0.0 2 0 2 1 54 1 2 120 258 0 0 147 0 0.4 1 0 3 1 51 1 2 94 227 0 1 154 1 0.0 2 1 3 1 29 1 1 130 204 0 0 202 0 0.0 2 0 2 1 51 1 0 140 261 0 0 186 1 0.0 2 0 2 1 43 0 2 122 213 0 1 165 0 0.2 1 0 2 1 55 0 1 135 250 0 0 161 0 1.4 1 0 2 1 51 1 2 125 245 1 0 166 0 2.4 1 0 2 1 59 1 1 140 221 0 1 164 1 0.0 2 0 2 1 52 1 1 128 205 1 1 184 0 0.0 2 0 2 1 58 1 2 105 240 0 0 154 1 0.6 1 0 3 1 41 1 2 112 250 0 1 179 0 0.0 2 0 2 1 45 1 1 128 308 0 0 170 0 0.0 2 0 2 1 60 0 2 102 318 0 1 160 0 0.0 2 1 2 1 52 1 3 152 298 1 1 178 0 1.2 1 0 3 1 42 0 0 102 265 0 0 122 0 0.6 1 0 2 1 67 0 2 115 564 0 0 160 0 1.6 1 0 3 1 68 1 2 118 277 0 1 151 0 1.0 2 1 3 1 46 1 1 101 197 1 1 156 0 0.0 2 0 3 1 54 0 2 110 214 0 1 158 0 1.6 1 0 2 1 58 0 0 100 248 0 0 122 0 1.0 1 0 2 1 48 1 2 124 255 1 1 175 0 0.0 2 2 2 1 57 1 0 132 207 0 1 168 1 0.0 2 0 3 1 52 1 2 138 223 0 1 169 0 0.0 2 4 2 1 54 0 1 132 288 1 0 159 1 0.0 2 1 2 1 45 0 1 112 160 0 1 138 0 0.0 1 0 2 1 53 1 0 142 226 0 0 111 1 0.0 2 0 3 1 62 0 0 140 394 0 0 157 0 1.2 1 0 2 1 52 1 0 108 233 1 1 147 0 0.1 2 3 3 1 43 1 2 130 315 0 1 162 0 1.9 2 1 2 1 53 1 2 130 246 1 0 173 0 0.0 2 3 2 1 42 1 3 148 244 0 0 178 0 0.8 2 2 2 1 59 1 3 178 270 0 0 145 0 4.2 0 0 3 1 63 0 1 140 195 0 1 179 0 0.0 2 2 2 1 42 1 2 120 240 1 1 194 0 0.8 0 0 3 1 50 1 2 129 196 0 1 163 0 0.0 2 0 2 1 68 0 2 120 211 0 0 115 0 1.5 1 0 2 1

#Exploratory Data Analysis
(mostly graphical)

In [0]:
%sql

SELECT count(sex),
CASE
    WHEN sex == 1 THEN "Male"
    ELSE "Female"
END AS sex
FROM HeartData group by sex;

count(sex) sex 207 Male 96 Female

In [0]:
%sql

SELECT count(cp),
CASE
    WHEN cp == 0 THEN "Typical angina"
    WHEN cp == 1 THEN "Atypical angina"
    WHEN cp == 2 THEN "Non-anginal pain"
    ELSE "Asymptomatic"
END AS ChestPain
FROM HeartData group by cp;

count(cp) ChestPain 50 Atypical angina 23 Asymptomatic 87 Non-anginal pain 143 Typical angina

In [0]:
%sql

SELECT count(fbs),
CASE
    WHEN fbs == 1 THEN "True"
    ELSE "False"
END AS FastingBloodSugar
FROM HeartData group by fbs;

count(fbs) FastingBloodSugar 45 True 258 False

In [0]:
%sql

SELECT count(restecg),
CASE
    WHEN restecg == 0 THEN "Normal"
    WHEN restecg == 1 THEN "Abnormality"
    ELSE "Left Ventricular Hypertrophy"
END AS RestingElectrocardioGraphic
FROM HeartData group by restecg;

count(restecg) RestingElectrocardioGraphic 152 Abnormality 4 Left Ventricular Hypertrophy 147 Normal

In [0]:
%sql

SELECT count(exang),
CASE
    WHEN exang == 0 THEN "Not Induced"
    ELSE "Induced"
END AS ExerciseInducedAngina
FROM HeartData group by exang;


count(exang) ExerciseInducedAngina 99 Induced 204 Not Induced

In [0]:
%sql

SELECT count(slope),
CASE
    WHEN slope == 0 THEN "Upsloping"
    WHEN slope == 1 THEN "Flat"
    ELSE "Downsloping"
END AS Slope
FROM HeartData group by slope;

count(slope) Slope 140 Flat 142 Downsloping 21 Upsloping

In [0]:
%sql

SELECT count(ca), ca FROM HeartData group by ca order by ca;

count(ca) ca 175 0 65 1 38 2 20 3 5 4

In [0]:
%sql

select thal, count(thal) FROM HeartData group by thal order by thal;

thal count(thal) 0 2 1 18 2 166 3 117

Output can only be rendered in Databricks

Output can only be rendered in Databricks

In [0]:
%sql

select count(sex),
CASE
    WHEN sex == 0 THEN "Female"
    ELSE "Male"
END AS Sex,
count(target), 
CASE
    WHEN target == 0 THEN "No Disease"
    ELSE "Has Disease"
  END AS DiseaseStatus
from HeartData group by sex, target;

count(sex) Sex count(target) DiseaseStatus 114 Male 114 No Disease 93 Male 93 Has Disease 24 Female 24 No Disease 72 Female 72 Has Disease

In [0]:
%sql

select count(cp),
CASE
    WHEN cp == 0 THEN "Typical angina"
    WHEN cp == 1 THEN "Atypical angina"
    WHEN cp == 2 THEN "Non-anginal pain"
    ELSE "Asymptomatic"
END AS ChestPain,
count(target), 
CASE
    WHEN target == 0 THEN "No Disease"
    ELSE "Has Disease"
  END AS DiseaseStatus
from HeartData group by cp, target order by cp;

count(cp) ChestPain count(target) DiseaseStatus 104 Typical angina 104 No Disease 39 Typical angina 39 Has Disease 9 Atypical angina 9 No Disease 41 Atypical angina 41 Has Disease 69 Non-anginal pain 69 Has Disease 18 Non-anginal pain 18 No Disease 16 Asymptomatic 16 Has Disease 7 Asymptomatic 7 No Disease

In [0]:
%sql

select count(fbs),
CASE
    WHEN fbs == 1 THEN "True"
    ELSE "False"
END AS FastingBloodSugar,
count(target), 
CASE
    WHEN target == 0 THEN "No Disease"
    ELSE "Has Disease"
  END AS DiseaseStatus
from HeartData group by fbs, target order by fbs;

count(fbs) FastingBloodSugar count(target) DiseaseStatus 116 False 116 No Disease 142 False 142 Has Disease 22 True 22 No Disease 23 True 23 Has Disease

In [0]:
%sql

SELECT count(restecg),
CASE
    WHEN restecg == 0 THEN "Normal"
    WHEN restecg == 1 THEN "Abnormality"
    ELSE "Left Ventricular Hypertrophy"
END AS RestingElectrocardioGraphic,
count(target), 
CASE
    WHEN target == 0 THEN "No Disease"
    ELSE "Has Disease"
  END AS DiseaseStatus
from HeartData group by restecg, target order by restecg;

count(restecg) RestingElectrocardioGraphic count(target) DiseaseStatus 79 Normal 79 No Disease 68 Normal 68 Has Disease 56 Abnormality 56 No Disease 96 Abnormality 96 Has Disease 1 Left Ventricular Hypertrophy 1 Has Disease 3 Left Ventricular Hypertrophy 3 No Disease

In [0]:
%sql

SELECT count(exang),
CASE
    WHEN exang == 0 THEN "Not Induced"
    ELSE "Induced"
END AS ExerciseInducedAngina,
count(target), 
CASE
    WHEN target == 0 THEN "No Disease"
    ELSE "Has Disease"
  END AS DiseaseStatus
from HeartData group by exang, target order by exang;

count(exang) ExerciseInducedAngina count(target) DiseaseStatus 62 Not Induced 62 No Disease 142 Not Induced 142 Has Disease 76 Induced 76 No Disease 23 Induced 23 Has Disease

In [0]:
%sql

SELECT count(slope),
CASE
    WHEN slope == 0 THEN "Upsloping"
    WHEN slope == 1 THEN "Flat"
    ELSE "Downsloping"
END AS Slope,
count(target), 
CASE
    WHEN target == 0 THEN "No Disease"
    ELSE "Has Disease"
  END AS DiseaseStatus
from HeartData group by slope, target order by slope;

count(slope) Slope count(target) DiseaseStatus 107 Downsloping 107 Has Disease 35 Downsloping 35 No Disease 91 Flat 91 No Disease 49 Flat 49 Has Disease 12 Upsloping 12 No Disease 9 Upsloping 9 Has Disease

In [0]:
%sql

SELECT ca, count(ca),
count(target), 
CASE
    WHEN target == 0 THEN "No Disease"
    ELSE "Has Disease"
  END AS DiseaseStatus
from HeartData group by ca, target order by ca;

ca count(ca) count(target) DiseaseStatus 0 45 45 No Disease 0 130 130 Has Disease 1 44 44 No Disease 1 21 21 Has Disease 2 7 7 Has Disease 2 31 31 No Disease 3 3 3 Has Disease 3 17 17 No Disease 4 1 1 No Disease 4 4 4 Has Disease

In [0]:
%sql

SELECT thal, count(thal),count(target), 
CASE
    WHEN target == 0 THEN "No Disease"
    ELSE "Has Disease"
  END AS DiseaseStatus
from HeartData group by thal, target order by thal;

thal count(thal) count(target) DiseaseStatus 0 1 1 No Disease 0 1 1 Has Disease 1 12 12 No Disease 1 6 6 Has Disease 2 130 130 Has Disease 2 36 36 No Disease 3 28 28 Has Disease 3 89 89 No Disease

In [0]:
%sql

select age from HeartData;

age 63 37 41 56 57 57 56 44 52 57 54 48 49 64 58 50 58 66 43 69 59 44 42 61 40 71 59 51 65 53 41 65 44 54 51 46 54 54 65 65 51 48 45 53 39 52 44 47 53 53 51 66 62 44 63 52 48 45 34 57 71 54 52 41 58 35 51 45 44 62 54 51 29 51 43 55 51 59 52 58 41 45 60 52 42 67 68 46 54 58 48 57 52 54 45 53 62 52 43 53 42 59 63 42 50 68 69 45 50 50 64 57 64 43 55 37 41 56 46 46 64 59 41 54 39 34 47 67 52 74 54 49 42 41 41 49 60 62 57 64 51 43 42 67 76 70 44 60 44 42 66 71 64 66 39 58 47 35 58 56 56 55 41 38 38 67 67 62 63 53 56 48 58 58 60 40 60 64 43 57 55 65 61 58 50 44 60 54 50 41 51 58 54 60 60 59 46 67 62 65 44 60 58 68 62 52 59 60 49 59 57 61 39 61 56 43 62 63 65 48 63 55 65 56 54 70 62 35 59 64 47 57 55 64 70 51 58 60 77 35 70 59 64 57 56 48 56 66 54 69 51 43 62 67 59 45 58 50 62 38 66 52 53 63 54 66 55 49 54 56 46 61 67 58 47 52 58 57 58 61 42 52 59 40 61 46 59 57 57 55 61 58 58 67 44 63 63 59 57 45 68 57 57

In [0]:
%sql

select trestbps from HeartData;

trestbps 145 130 130 120 120 140 140 120 172 150 140 130 130 110 150 120 120 150 150 140 135 130 140 150 140 160 150 110 140 130 105 120 130 125 125 142 135 150 155 160 140 130 104 130 140 120 140 138 128 138 130 120 130 108 135 134 122 115 118 128 110 108 118 135 140 138 100 130 120 124 120 94 130 140 122 135 125 140 128 105 112 128 102 152 102 115 118 101 110 100 124 132 138 132 112 142 140 108 130 130 148 178 140 120 129 120 160 138 120 110 180 150 140 110 130 120 130 120 105 138 130 138 112 108 94 118 112 152 136 120 160 134 120 110 126 130 120 128 110 128 120 115 120 106 140 156 118 150 120 130 160 112 170 146 138 130 130 122 125 130 120 132 120 138 138 160 120 140 130 140 130 110 120 132 130 110 117 140 120 150 132 150 130 112 150 112 130 124 140 110 130 128 120 145 140 170 150 125 120 110 110 125 150 180 160 128 110 150 120 140 128 120 118 145 125 132 130 130 135 130 150 140 138 200 110 145 120 120 170 125 108 165 160 120 130 140 125 140 125 126 160 174 145 152 132 124 134 160 192 140 140 132 138 100 160 142 128 144 150 120 178 112 123 108 110 112 180 118 122 130 120 134 120 100 110 125 146 124 136 138 136 128 126 152 140 140 134 154 110 128 148 114 170 152 120 140 124 164 140 110 144 130 130

In [0]:
%sql

select chol from HeartData;

chol 233 250 204 236 354 192 294 263 199 168 239 275 266 211 283 219 340 226 247 239 234 233 226 243 199 302 212 175 417 197 198 177 219 273 213 177 304 232 269 360 308 245 208 264 321 325 235 257 216 234 256 302 231 141 252 201 222 260 182 303 265 309 186 203 211 183 222 234 220 209 258 227 204 261 213 250 245 221 205 240 250 308 318 298 265 564 277 197 214 248 255 207 223 288 160 226 394 233 315 246 244 270 195 240 196 211 234 236 244 254 325 126 313 211 262 215 214 193 204 243 303 271 268 267 199 210 204 277 196 269 201 271 295 235 306 269 178 208 201 263 295 303 209 223 197 245 242 240 226 180 228 149 227 278 220 197 253 192 220 221 240 342 157 175 175 286 229 268 254 203 256 229 284 224 206 167 230 335 177 276 353 225 330 230 243 290 253 266 233 172 305 216 188 282 185 326 231 254 267 248 197 258 270 274 164 255 239 258 188 177 229 260 219 307 249 341 263 330 254 256 407 217 282 288 239 174 281 198 288 309 243 289 289 246 322 299 300 293 304 282 269 249 212 274 184 274 409 246 283 254 298 247 294 299 273 309 259 200 244 231 228 230 282 269 206 212 327 149 286 283 249 234 237 234 275 212 218 261 319 166 315 204 218 223 207 311 204 232 335 205 203 318 225 212 169 187 197 176 241 264 193 131 236

In [0]:
%sql

select thalach from HeartData;

thalach 150 187 172 178 163 148 153 173 162 174 160 139 171 144 162 158 172 114 171 151 161 179 178 137 178 162 157 123 157 152 168 140 188 152 125 160 170 165 148 151 142 180 148 143 182 172 180 156 115 160 149 151 146 175 172 158 186 185 174 159 130 156 190 132 165 182 143 175 170 163 147 154 202 186 165 161 166 164 184 154 179 170 160 178 122 160 151 156 158 122 175 168 169 159 138 111 157 147 162 173 178 145 179 194 163 115 131 152 162 159 154 173 133 161 155 170 168 162 172 152 122 182 172 167 179 192 143 172 169 121 163 162 162 153 163 163 96 140 126 105 157 181 173 142 116 143 149 171 169 150 138 125 155 152 152 131 179 174 144 163 169 166 182 173 173 108 129 160 147 155 142 168 160 173 132 114 160 158 120 112 132 114 169 165 128 153 144 109 163 158 142 131 113 142 155 140 147 163 99 158 177 141 111 150 145 161 142 157 139 162 150 140 140 146 144 136 97 132 127 150 154 111 174 133 126 125 103 130 159 131 152 124 145 96 109 173 171 170 162 156 112 143 132 88 105 166 150 120 195 146 122 143 106 125 125 147 130 126 154 182 165 160 95 169 108 132 117 126 116 103 144 145 71 156 118 168 105 141 152 125 125 156 134 181 138 120 162 164 143 130 161 140 146 150 144 144 136 90 123 132 141 115 174

In [0]:
%sql

select oldpeak from HeartData;

oldpeak 2.3 3.5 1.4 0.8 0.6 0.4 1.3 0.0 0.5 1.6 1.2 0.2 0.6 1.8 1.0 1.6 0.0 2.6 1.5 1.8 0.5 0.4 0.0 1.0 1.4 0.4 1.6 0.6 0.8 1.2 0.0 0.4 0.0 0.5 1.4 1.4 0.0 1.6 0.8 0.8 1.5 0.2 3.0 0.4 0.0 0.2 0.0 0.0 0.0 0.0 0.5 0.4 1.8 0.6 0.0 0.8 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.4 1.2 0.6 0.0 0.0 0.4 0.0 0.0 0.0 0.2 1.4 2.4 0.0 0.0 0.6 0.0 0.0 0.0 1.2 0.6 1.6 1.0 0.0 1.6 1.0 0.0 0.0 0.0 0.0 0.0 0.0 1.2 0.1 1.9 0.0 0.8 4.2 0.0 0.8 0.0 1.5 0.1 0.2 1.1 0.0 0.0 0.2 0.2 0.0 0.0 0.0 2.0 1.9 0.0 0.0 2.0 0.0 0.0 0.0 0.0 0.7 0.1 0.0 0.1 0.2 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.5 0.2 0.6 1.2 0.0 0.3 1.1 0.0 0.3 0.9 0.0 0.0 2.3 1.6 0.6 0.0 0.0 0.6 0.0 0.0 0.4 0.0 0.0 1.2 0.0 0.0 0.0 1.5 2.6 3.6 1.4 3.1 0.6 1.0 1.8 3.2 2.4 2.0 1.4 0.0 2.5 0.6 1.2 1.0 0.0 2.5 2.6 0.0 1.4 2.2 0.6 0.0 1.2 2.2 1.4 2.8 3.0 3.4 3.6 0.2 1.8 0.6 0.0 2.8 0.8 1.6 6.2 0.0 1.2 2.6 2.0 0.0 0.4 3.6 1.2 1.0 1.2 3.0 1.2 1.8 2.8 0.0 4.0 5.6 1.4 4.0 2.8 2.6 1.4 1.6 0.2 1.8 0.0 1.0 0.8 2.2 2.4 1.6 0.0 1.2 0.0 0.0 2.9 0.0 2.0 1.2 2.1 0.5 1.9 0.0 0.0 2.0 4.2 0.1 1.9 0.9 0.0 0.0 3.0 0.9 1.4 3.8 1.0 0.0 2.0 1.8 0.0 0.1 3.4 0.8 3.2 1.6 0.8 2.6 1.0 0.1 1.0 1.0 2.0 0.3 0.0 3.6 1.8 1.0 2.2 0.0 1.9 1.8 0.8 0.0 3.0 2.0 0.0 4.4 2.8 0.8 2.8 4.0 0.0 1.0 0.2 1.2 3.4 1.2 0.0

In [0]:
%sql

select target from HeartData;

target 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0

In [0]:
%sql

select age,count(age) as AgeCounter from HeartData group by age order by age;

age AgeCounter 29 1 34 2 35 4 37 2 38 3 39 4 40 3 41 10 42 8 43 8 44 11 45 8 46 7 47 5 48 7 49 5 50 7 51 12 52 13 53 8 54 16 55 8 56 11 57 17 58 19 59 14 60 11 61 8 62 11 63 9 64 10 65 8 66 7 67 9 68 4 69 3 70 4 71 3 74 1 76 1 77 1

In [0]:
%sql

SELECT 
CASE
    WHEN age >=29 and age <40 THEN "Young Ages"
    WHEN age >=40 and age <55 THEN "Middle Ages"
    ELSE "Elderly Ages"
  END AS Age,
  count(Age)
from HeartData group by Age;

Age count(Age) Elderly Ages 8 Middle Ages 8 Young Ages 2 Elderly Ages 1 Middle Ages 11 Middle Ages 5 Middle Ages 13 Middle Ages 3 Elderly Ages 17 Middle Ages 16 Middle Ages 7 Elderly Ages 10 Middle Ages 10 Middle Ages 8 Young Ages 2 Elderly Ages 8 Young Ages 4 Elderly Ages 14 Elderly Ages 8 Young Ages 4 Middle Ages 5 Middle Ages 12 Elderly Ages 3 Elderly Ages 9 Elderly Ages 1 Middle Ages 7 Middle Ages 8 Young Ages 3 Elderly Ages 4 Elderly Ages 11 Young Ages 1 Elderly Ages 11 Elderly Ages 11 Elderly Ages 19 Elderly Ages 3 Elderly Ages 4 Middle Ages 8 Elderly Ages 7 Middle Ages 7 Elderly Ages 9 Elderly Ages 1

In [0]:
%sql

SELECT 
CASE
    WHEN age >=29 and age <40 THEN "Young Ages"
    WHEN age >=40 and age <55 THEN "Middle Ages"
    ELSE "Elderly Ages"
  END AS Age,
CASE
    WHEN sex == 1 THEN "Male"
    ELSE "Female"
END AS sex
from HeartData;

Age sex Elderly Ages Male Young Ages Male Middle Ages Female Elderly Ages Male Elderly Ages Female Elderly Ages Male Elderly Ages Female Middle Ages Male Middle Ages Male Elderly Ages Male Middle Ages Male Middle Ages Female Middle Ages Male Elderly Ages Male Elderly Ages Female Middle Ages Female Elderly Ages Female Elderly Ages Female Middle Ages Male Elderly Ages Female Elderly Ages Male Middle Ages Male Middle Ages Male Elderly Ages Male Middle Ages Male Elderly Ages Female Elderly Ages Male Middle Ages Male Elderly Ages Female Middle Ages Male Middle Ages Female Elderly Ages Male Middle Ages Male Middle Ages Male Middle Ages Male Middle Ages Female Middle Ages Female Middle Ages Male Elderly Ages Female Elderly Ages Female Middle Ages Female Middle Ages Male Middle Ages Male Middle Ages Female Young Ages Male Middle Ages Male Middle Ages Male Middle Ages Male Middle Ages Female Middle Ages Female Middle Ages Female Elderly Ages Male Elderly Ages Male Middle Ages Female Elderly Ages Female Middle Ages Male Middle Ages Male Middle Ages Male Young Ages Male Elderly Ages Female Elderly Ages Female Middle Ages Male Middle Ages Male Middle Ages Male Elderly Ages Male Young Ages Female Middle Ages Male Middle Ages Female Middle Ages Male Elderly Ages Female Middle Ages Male Middle Ages Male Young Ages Male Middle Ages Male Middle Ages Female Elderly Ages Female Middle Ages Male Elderly Ages Male Middle Ages Male Elderly Ages Male Middle Ages Male Middle Ages Male Elderly Ages Female Middle Ages Male Middle Ages Female Elderly Ages Female Elderly Ages Male Middle Ages Male Middle Ages Female Elderly Ages Female Middle Ages Male Elderly Ages Male Middle Ages Male Middle Ages Female Middle Ages Female Middle Ages Male Elderly Ages Female Middle Ages Male Middle Ages Male Middle Ages Male Middle Ages Male Elderly Ages Male Elderly Ages Female Middle Ages Male Middle Ages Male Elderly Ages Female Elderly Ages Male Middle Ages Female Middle Ages Female Middle Ages Female Elderly Ages Female Elderly Ages Male Elderly Ages Female Middle Ages Male Elderly Ages Male Young Ages Female Middle Ages Male Elderly Ages Male Middle Ages Female Middle Ages Female Elderly Ages Female Elderly Ages Male Middle Ages Female Middle Ages Female Young Ages Female Young Ages Female Middle Ages Male Elderly Ages Female Middle Ages Female Elderly Ages Female Middle Ages Female Middle Ages Female Middle Ages Male Middle Ages Male Middle Ages Female Middle Ages Female Elderly Ages Female Elderly Ages Male Elderly Ages Male Elderly Ages Male Middle Ages Female Middle Ages Male Middle Ages Female Elderly Ages Female Elderly Ages Female Elderly Ages Male Middle Ages Female Elderly Ages Female Middle Ages Male Middle Ages Male Elderly Ages Male Elderly Ages Female Elderly Ages Male Elderly Ages Female Young Ages Female Elderly Ages Female Middle Ages Male Young Ages Male Elderly Ages Male Elderly Ages Male Elderly Ages Male Elderly Ages Female Middle Ages Male Young Ages Male Young Ages Male Elderly Ages Male Elderly Ages Male Elderly Ages Female Elderly Ages Male Middle Ages Male Elderly Ages Male Middle Ages Male Elderly Ages Male Elderly Ages Male Elderly Ages Male Middle Ages Male Elderly Ages Male Elderly Ages Male Middle Ages Male Elderly Ages Male Elderly Ages Male Elderly Ages Female Elderly Ages Female Elderly Ages Male Middle Ages Male Middle Ages Male Elderly Ages Male Middle Ages Male Middle Ages Male Middle Ages Male Middle Ages Female Elderly Ages Male Middle Ages Male Elderly Ages Male Elderly Ages Male Elderly Ages Male Middle Ages Male Elderly Ages Male Elderly Ages Male Elderly Ages Male Middle Ages Male Elderly Ages Male Elderly Ages Male Elderly Ages Male Elderly Ages Female Middle Ages Male Elderly Ages Male Elderly Ages Female Middle Ages Male Elderly Ages Male Elderly Ages Male Elderly Ages Male Young Ages Male Elderly Ages Female Elderly Ages Male Middle Ages Female Elderly Ages Female Elderly Ages Male Elderly Ages Male Middle Ages Male E

In [0]:
%sql

select trestbps, age from HeartData;

trestbps age 145 63 130 37 130 41 120 56 120 57 140 57 140 56 120 44 172 52 150 57 140 54 130 48 130 49 110 64 150 58 120 50 120 58 150 66 150 43 140 69 135 59 130 44 140 42 150 61 140 40 160 71 150 59 110 51 140 65 130 53 105 41 120 65 130 44 125 54 125 51 142 46 135 54 150 54 155 65 160 65 140 51 130 48 104 45 130 53 140 39 120 52 140 44 138 47 128 53 138 53 130 51 120 66 130 62 108 44 135 63 134 52 122 48 115 45 118 34 128 57 110 71 108 54 118 52 135 41 140 58 138 35 100 51 130 45 120 44 124 62 120 54 94 51 130 29 140 51 122 43 135 55 125 51 140 59 128 52 105 58 112 41 128 45 102 60 152 52 102 42 115 67 118 68 101 46 110 54 100 58 124 48 132 57 138 52 132 54 112 45 142 53 140 62 108 52 130 43 130 53 148 42 178 59 140 63 120 42 129 50 120 68 160 69 138 45 120 50 110 50 180 64 150 57 140 64 110 43 130 55 120 37 130 41 120 56 105 46 138 46 130 64 138 59 112 41 108 54 94 39 118 34 112 47 152 67 136 52 120 74 160 54 134 49 120 42 110 41 126 41 130 49 120 60 128 62 110 57 128 64 120 51 115 43 120 42 106 67 140 76 156 70 118 44 150 60 120 44 130 42 160 66 112 71 170 64 146 66 138 39 130 58 130 47 122 35 125 58 130 56 120 56 132 55 120 41 138 38 138 38 160 67 120 67 140 62 130 63 140 53 130 56 110 48 120 58 132 58 130 60 110 40 117 60 140 64 120 43 150 57 132 55 150 65 130 61 112 58 150 50 112 44 130 60 124 54 140 50 110 41 130 51 128 58 120 54 145 60 140 60 170 59 150 46 125 67 120 62 110 65 110 44 125 60 150 58 180 68 160 62 128 52 110 59 150 60 120 49 140 59 128 57 120 61 118 39 145 61 125 56 132 43 130 62 130 63 135 65 130 48 150 63 140 55 138 65 200 56 110 54 145 70 120 62 120 35 170 59 125 64 108 47 165 57 160 55 120 64 130 70 140 51 125 58 140 60 125 77 126 35 160 70 174 59 145 64 152 57 132 56 124 48 134 56 160 66 192 54 140 69 140 51 132 43 138 62 100 67 160 59 142 45 128 58 144 50 150 62 120 38 178 66 112 52 123 53 108 63 110 54 112 66 180 55 118 49 122 54 130 56 120 46 134 61 120 67 100 58 110 47 125 52 146 58 124 57 136 58 138 61 136 42 128 52 126 59 152 40 140 61 140 46 134 59 154 57 110 57 128 55 148 61 114 58 170 58 152 67 120 44 140 63 124 63 164 59 140 57 110 45 144 68 130 57 130 57

In [0]:
%sql

select age, thalach from HeartData order by age asc; 

age thalach 29 202 34 174 34 192 35 182 35 174 35 130 35 156 37 187 37 170 38 173 38 173 38 182 39 182 39 179 39 152 39 140 40 178 40 114 40 181 41 172 41 168 41 132 41 179 41 168 41 172 41 153 41 163 41 182 41 158 42 178 42 122 42 178 42 194 42 162 42 173 42 150 42 125 43 171 43 165 43 162 43 161 43 181 43 120 43 136 43 143 44 173 44 179 44 188 44 180 44 175 44 170 44 149 44 169 44 153 44 177 44 144 45 148 45 185 45 175 45 170 45 138 45 152 45 147 45 132 46 160 46 156 46 172 46 152 46 147 46 144 46 120 47 156 47 143 47 179 47 152 47 118 48 139 48 180 48 186 48 175 48 168 48 150 48 166 49 171 49 162 49 163 49 139 49 126 50 158 50 163 50 162 50 159 50 128 50 163 50 126 51 123 51 125 51 142 51 149 51 143 51 154 51 186 51 166 51 157 51 142 51 173 51 122 52 162 52 172 52 158 52 190 52 184 52 178 52 169 52 147 52 169 52 161 52 160 52 168 52 156 53 152 53 143 53 115 53 160 53 111 53 173 53 155 53 95 54 160 54 152 54 170 54 165 54 156 54 147 54 158 54 159 54 167 54 163 54 109 54 113 54 126 54 195 54 108 54 116 55 161 55 155 55 166 55 132 55 111 55 145 55 117 55 130 56 178 56 153 56 162 56 163 56 169 56 142 56 144 56 133 56 105 56 150 56 103 57 163 57 148 57 174 57 159 57 168 57 173 57 126 57 112 57 150 57 124 57 88 57 141 57 164 57 143 57 123 57 115 57 174 58 162 58 172 58 165 58 154 58 122 58 131 58 144 58 160 58 173 58 165 58 131 58 111 58 171 58 130 58 156 58 105 58 152 58 140 58 146 59 161 59 157 59 164 59 145 59 182 59 140 59 142 59 162 59 159 59 143 59 125 59 134 59 162 59 90 60 160 60 96 60 171 60 132 60 160 60 144 60 142 60 155 60 141 60 157 60 170 61 137 61 169 61 140 61 146 61 145 61 125 61 138 61 161 62 146 62 163 62 157 62 140 62 160 62 99 62 145 62 97 62 103 62 106 62 154 63 150 63 172 63 179 63 147 63 132 63 154 63 169 63 144 63 136 64 144 64 154 64 133 64 122 64 105 64 155 64 158 64 131 64 96 64 132 65 157 65 140 65 148 65 151 65 114 65 158 65 127 65 174 66 114 66 151 66 138 66 152 66 120 66 165 66 132 67 160 67 172 67 142 67 108 67 129 67 163 67 125 67 71 67 150 68 151 68 115 68 150 68 141 69 151 69 131 69 146 70 143 70 125 70 109 70 112 71 162 71 130 71 125 74 121 76 116 77 162

### Classification Model

A decision tree classifier that uses features of patient details to predict the class label(Yes or No).

In [0]:
%scala

import org.apache.spark.sql.types._
import org.apache.spark.sql.functions._

import org.apache.spark.ml.classification.DecisionTreeClassificationModel
import org.apache.spark.ml.classification.DecisionTreeClassifier
import org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator
import org.apache.spark.ml.feature.VectorAssembler

import org.apache.spark.sql.types._
import org.apache.spark.sql.functions._
import org.apache.spark.ml.classification.DecisionTreeClassificationModel
import org.apache.spark.ml.classification.DecisionTreeClassifier
import org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator
import org.apache.spark.ml.feature.VectorAssembler

### Prepare the Training Data
We'll use the **VectorAssembler** class to transform the feature columns into a vector, and then rename the **Attrition** column to **label**.

###VectorAssembler()

 * VectorAssembler():  is a transformer that combines a given list of columns into a single vector column.
 * It is useful for combining raw features and features generated by different feature transformers into a single feature vector, in order to train ML models like logistic regression and decision trees. 

* VectorAssembler accepts the following input column types -> all numeric types, boolean types, and vector types.

* In each row, the values of the input columns will be concatenated into a vector in the specified order.

### Split the Data
It is common practice when building machine learning models to split the source data, using some of it to train the model and reserving some to test the trained model. 70% of the data for training, and 30% for testing.

In [0]:
%scala

val splits = heartDF.randomSplit(Array(0.7, 0.3))
val train = splits(0)
val test = splits(1)
val train_rows = train.count()
val test_rows = test.count()
println("Training Rows-: " + train_rows + " Testing Rows- " + test_rows)

Training Rows: 203 Testing Rows: 100
splits: Array[org.apache.spark.sql.Dataset[org.apache.spark.sql.Row]] = Array([age: int, sex: int ... 12 more fields], [age: int, sex: int ... 12 more fields])
train: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [age: int, sex: int ... 12 more fields]
test: org.apache.spark.sql.Dataset[org.apache.spark.sql.Row] = [age: int, sex: int ... 12 more fields]
train_rows: Long = 203
test_rows: Long = 100

In [0]:
%scala

import org.apache.spark.ml.feature.VectorAssembler

val assembler = new VectorAssembler().setInputCols(Array("age", "sex", "cp", "trestbps", "chol", "fbs", "restecg", "thalach", "exang", "oldpeak", "slope", "ca", "thal")).setOutputCol("features")
// all the columns except target

val training = assembler.transform(train).select($"features", $"target".alias("label"))

training.show()

+--------------------+-----+
 features|label|
+--------------------+-----+
[34.0,1.0,3.0,118...| 1|
[35.0,1.0,0.0,120...| 0|
[35.0,1.0,0.0,126...| 0|
[35.0,1.0,1.0,122...| 1|
[37.0,0.0,2.0,120...| 1|
[37.0,1.0,2.0,130...| 1|
[38.0,1.0,2.0,138...| 1|
[39.0,0.0,2.0,94....| 1|
[39.0,1.0,2.0,140...| 1|
[40.0,1.0,0.0,110...| 0|
[40.0,1.0,3.0,140...| 1|
[41.0,0.0,1.0,105...| 1|
[41.0,0.0,2.0,112...| 1|
[41.0,1.0,1.0,110...| 1|
[41.0,1.0,1.0,120...| 1|
[41.0,1.0,1.0,135...| 1|
(13,[0,3,4,7,9,10...| 1|
[42.0,0.0,2.0,120...| 1|
[42.0,1.0,0.0,140...| 1|
[42.0,1.0,1.0,120...| 1|
+--------------------+-----+
only showing top 20 rows

import org.apache.spark.ml.feature.VectorAssembler
assembler: org.apache.spark.ml.feature.VectorAssembler = VectorAssembler: uid=vecAssembler_7e05e43b50f3, handleInvalid=error, numInputCols=13
training: org.apache.spark.sql.DataFrame = [features: vector, label: int]

### Train a Classification Model (Decision tree classifier)
Next, you need to train a Classification Model using the training data. To do this, create an instance of the Decision tree classifier algorithm you want to use and use its **fit** method to train a model based on the training DataFrame. In this Project, you will use a *Decision tree classifier* algorithm

In [0]:
%scala

import org.apache.spark.ml.classification.DecisionTreeClassificationModel
import org.apache.spark.ml.classification.DecisionTreeClassifier
import org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator

val dt = new DecisionTreeClassifier().setLabelCol("label").setFeaturesCol("features")

val model = dt.fit(training)

println("Model Trained.")

Model Trained.
import org.apache.spark.ml.classification.DecisionTreeClassificationModel
import org.apache.spark.ml.classification.DecisionTreeClassifier
import org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator
dt: org.apache.spark.ml.classification.DecisionTreeClassifier = dtc_b8101958e43d
model: org.apache.spark.ml.classification.DecisionTreeClassificationModel = DecisionTreeClassificationModel: uid=dtc_b8101958e43d, depth=5, numNodes=37, numClasses=2, numFeatures=13

### Prepare the Testing Data
**target** column -> **newLabel**.

In [0]:
%scala

val testing = assembler.transform(test).select($"features", $"target".alias("newLabel"))
testing.show()

+--------------------+--------+
 features|newLabel|
+--------------------+--------+
[29.0,1.0,1.0,130...| 1|
[34.0,0.0,1.0,118...| 1|
[35.0,0.0,0.0,138...| 1|
[38.0,1.0,2.0,138...| 1|
[38.0,1.0,3.0,120...| 0|
[39.0,0.0,2.0,138...| 1|
[39.0,1.0,0.0,118...| 0|
[40.0,1.0,0.0,152...| 0|
[41.0,0.0,1.0,126...| 1|
[41.0,0.0,1.0,130...| 1|
(13,[0,1,3,4,7,10...| 0|
[41.0,1.0,2.0,112...| 1|
[41.0,1.0,2.0,130...| 1|
[42.0,1.0,0.0,136...| 0|
[43.0,1.0,0.0,115...| 1|
[43.0,1.0,0.0,120...| 0|
[43.0,1.0,0.0,132...| 0|
[44.0,1.0,0.0,110...| 0|
[44.0,1.0,0.0,120...| 0|
[44.0,1.0,1.0,130...| 1|
+--------------------+--------+
only showing top 20 rows

testing: org.apache.spark.sql.DataFrame = [features: vector, newLabel: int]

### Test the Model
Time to make some predictions

In [0]:
%scala

val prediction = model.transform(testing)
val predicted = prediction.select("features", "prediction", "newLabel")
predicted.show(100)

+--------------------+----------+--------+
 features|prediction|newLabel|
+--------------------+----------+--------+
[29.0,1.0,1.0,130...| 1.0| 1|
[34.0,0.0,1.0,118...| 1.0| 1|
[35.0,0.0,0.0,138...| 1.0| 1|
[38.0,1.0,2.0,138...| 1.0| 1|
[38.0,1.0,3.0,120...| 1.0| 0|
[39.0,0.0,2.0,138...| 1.0| 1|
[39.0,1.0,0.0,118...| 0.0| 0|
[40.0,1.0,0.0,152...| 1.0| 0|
[41.0,0.0,1.0,126...| 1.0| 1|
[41.0,0.0,1.0,130...| 1.0| 1|
(13,[0,1,3,4,7,10...| 1.0| 0|
[41.0,1.0,2.0,112...| 1.0| 1|
[41.0,1.0,2.0,130...| 1.0| 1|
[42.0,1.0,0.0,136...| 1.0| 0|
[43.0,1.0,0.0,115...| 1.0| 1|
[43.0,1.0,0.0,120...| 0.0| 0|
[43.0,1.0,0.0,132...| 0.0| 0|
[44.0,1.0,0.0,110...| 0.0| 0|
[44.0,1.0,0.0,120...| 1.0| 0|
[44.0,1.0,1.0,130...| 1.0| 1|
[45.0,0.0,1.0,130...| 1.0| 1|
[46.0,0.0,1.0,105...| 1.0| 1|
[46.0,1.0,0.0,140...| 0.0| 0|
[46.0,1.0,2.0,150...| 0.0| 0|
[47.0,1.0,2.0,130...| 1.0| 1|
[48.0,0.0,2.0,130...| 1.0| 1|
[48.0,1.0,0.0,124...| 0.0| 0|
[48.0,1.0,0.0,130...| 0.0| 0|
[48.0,1.0,1.0,130...| 1.0| 1|
[48.0,1.0,2.0,124...| 0.0| 1|
[49.0,1.0,1.0,130...| 1.0| 1|
[50.0,0.0,1.0,120...| 1.0| 1|
[50.0,1.0,2.0,129...| 1.0| 1|
[51.0,0.0,0.0,130...| 0.0| 0|
[51.0,0.0,2.0,130...| 1.0| 1|
[51.0,0.0,2.0,140...| 0.0| 1|
[51.0,1.0,0.0,140...| 1.0| 1|
[51.0,1.0,2.0,94....| 0.0| 1|
[52.0,1.0,0.0,125...| 0.0| 0|
[52.0,1.0,1.0,128...| 1.0| 1|
[52.0,1.0,2.0,138...| 0.0| 1|
[52.0,1.0,3.0,118...| 1.0| 1|
[54.0,1.0,1.0,108...| 0.0| 1|
[54.0,1.0,2.0,150...| 1.0| 1|
[55.0,0.0,1.0,132...| 1.0| 1|
[55.0,0.0,1.0,135...| 1.0| 1|
[55.0,1.0,0.0,132...| 0.0| 0|
[55.0,1.0,0.0,140...| 0.0| 0|
[55.0,1.0,0.0,160...| 0.0| 0|
[56.0,1.0,0.0,130...| 0.0| 0|
[56.0,1.0,0.0,132...| 0.0| 0|
[56.0,1.0,1.0,120...| 1.0| 1|
[57.0,0.0,1.0,130...| 1.0| 0|
(13,[0,3,4,7,9,10...| 1.0| 1|
[58.0,0.0,0.0,130...| 1.0| 1|
[58.0,0.0,3.0,150...| 1.0| 1|
[58.0,1.0,0.0,125...| 0.0| 0|
[58.0,1.0,1.0,120...| 1.0| 0|
[58.0,1.0,1.0,125...| 0.0| 1|
[58.0,1.0,2.0,112...| 0.0| 0|
[59.0,0.0,0.0,174...| 1.0| 0|
[59.0,1.0,0.0,110...| 0.0| 0|
(13,[0,1,3,4,7,10...| 1.0| 1|
[59.0,1.0,0.0,140...| 0.0| 0|
[59.0,1.0,0.0,164...| 0.0| 0|
[59.0,1.0,0.0,170...| 0.0| 0|
[59.0,1.0,2.0,126...| 0.0| 0|
[59.0,1.0,2.0,150...| 1.0| 1|
[59.0,1.0,3.0,160...| 1.0| 0|
[59.0,1.0,3.0,178...| 0.0| 1|
[60.0,0.0,2.0,120...| 1.0| 1|
[60.0,1.0,0.0,145...| 0.0| 0|
(13,[0,3,4,7,10,1...| 1.0| 0|
[61.0,0.0,0.0,145...| 0.0| 0|
[61.0,1.0,0.0,120...| 0.0| 0|
[61.0,1.0,0.0,138...| 0.0| 0|
[61.0,1.0,2.0,150...| 1.0| 1|
[62.0,0.0,0.0,138...| 0.0| 0|
(13,[0,3,4,7,9,11...| 0.0| 0|
[62.0,0.0,2.0,130...| 1.0| 0|
[62.0,1.0,0.0,120...| 0.0| 0|
[62.0,1.0,2.0,130...| 0.0| 1|
(13,[0,2,3,4,7,10...| 1.0| 1|
[63.0,1.0,0.0,130...| 0.0| 0|
[64.0,0.0,0.0,130...| 0.0| 1|
[64.0,1.0,0.0,120...| 0.0| 0|
[64.0,1.0,0.0,145...| 0.0| 0|
[65.0,0.0,2.0,160...| 1.0| 1|
[65.0,1.0,0.0,110...| 0.0| 0|
[66.0,0.0,0.0,178...| 0.0| 0|
[66.0,0.0,3.0,150...| 0.0| 1|
[66.0,1.0,0.0,160...| 1.0| 1|
[66.0,1.0,1.0,160...| 0.0| 0|
[67.0,0.0,2.0,152...| 1.0| 1|
[67.0,1.0,0.0,120...| 1.0| 0|
[67.0,1.0,2.0,152...| 0.0| 0|
[69.0,0.0,3.0,140...| 1.0| 1|
[70.0,1.0,0.0,145...| 0.0| 0|
[70.0,1.0,2.0,160...| 0.0| 0|
[74.0,0.0,1.0,120...| 1.0| 1|
+--------------------+----------+--------+

prediction: org.apache.spark.sql.DataFrame = [features: vector, newLabel: int ... 3 more fields]
predicted: org.apache.spark.sql.DataFrame = [features: vector, prediction: double ... 1 more field]

The prediction column contains the predicted value for the label, and the newLabel column contains the actual known value from the testing data. We can observe that every prediction does not match with the actual value. These are residuals.

### Classification model Evalation

spark.mllib comes with a number of machine learning algorithms that can be used to learn from and make predictions on data. When these algorithms are applied to build machine learning models, there is a need to evaluate the performance of the model on some criteria, which depends on the application and its requirements. spark.mllib also provides a suite of metrics for the purpose of evaluating the performance of machine learning models.

In [0]:
%scala

val evaluator = new MulticlassClassificationEvaluator()
  .setLabelCol("newLabel")
  .setPredictionCol("prediction")
  .setMetricName("accuracy")
val accuracy = evaluator.evaluate(prediction)



evaluator: org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator = MulticlassClassificationEvaluator: uid=mcEval_222d1056420c, metricName=accuracy, metricLabel=0.0, beta=1.0, eps=1.0E-15
accuracy: Double = 0.78

In [0]:
%scala
The accuracy depends on how the data is split. For now, it is 78%

### Confusion Matrix Metrics

In [0]:
%scala

val tp = predicted.filter("prediction == 1 AND newlabel == 1").count().toFloat
val fp = predicted.filter("prediction == 1 AND newlabel == 0").count().toFloat
val tn = predicted.filter("prediction == 0 AND newlabel == 0").count().toFloat
val fn = predicted.filter("prediction == 0 AND newlabel == 1").count().toFloat
val metrics = spark.createDataFrame(Seq(
 ("TP", tp),
 ("FP", fp),
 ("TN", tn),
 ("FN", fn),
 ("Precision", tp / (tp + fp)),
 ("Recall", tp / (tp + fn)))).toDF("metric", "value")
metrics.show()

+---------+---------+
 metric| value|
+---------+---------+
 TP| 40.0|
 FP| 12.0|
 TN| 38.0|
 FN| 10.0|
Precision|0.7692308|
 Recall| 0.8|
+---------+---------+

tp: Float = 40.0
fp: Float = 12.0
tn: Float = 38.0
fn: Float = 10.0
metrics: org.apache.spark.sql.DataFrame = [metric: string, value: float]

This is how we can predict weather a person has heart disease or not.